clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -244,6 +258,53 @@ public void forceClose() {
         initiateClose();
     }
 
+    private void getPID(long remainingTime) {
+        // Send an InitPIDRequest, wait for the response, retrieve the PID from the response if success. Return it.
+        // Otherwise throw exception.
+        // Pick a random broker to send the initPIDRequest to.
+        metadata.requestUpdate();
+        long start = time.milliseconds();
+        Node node = client.leastLoadedNode(time.milliseconds());
+        while (!client.ready(node, time.milliseconds()) && 0 < remainingTime) {
+            if (client.connectionFailed(node)) {
+                node = client.leastLoadedNode(time.milliseconds());
+                client.ready(node, time.milliseconds());
+            }
+            client.poll(100, time.milliseconds());
```

Comment:
Instead of 100, maybe we could use the reconnect timeout? By the way, if we're checking for a disconnect, the time to do it is right after a call to `poll`.